### PR TITLE
Rewrite DemoteLocalTemporariesToVariablesPass

### PR DESCRIPTION
**Pull request by jdahm**
## Description

Cleans up the logic in the `DemoteLocalTemporariesToVariablesPass`.

### CODE REVIEW

File: src/gt4py/analysis/passes.py [hunk 1]
```diff
@@ -1195,30 +1199,40 @@ def apply(cls, node: gt_ir.StencilImplementation) -> Set[str]:
...
+            self.visit(node.target, **kwargs, is_target=True)
+
+        def visit_FieldRef(self, node: gt_ir.FieldRef, **kwargs) -> None:
+            # 1. is never used with an offset
+            if any(val != 0 for val in node.offset.values()):
+                self.demotables.discard(node.name)
+
+            # 2. is never read before assigned to
+            if kwargs.get("is_value", False) and node.name not in self.written_fields:
+                self.read_fields.add(node.name)
```
Comment (fthaler) [inline src/gt4py/analysis/passes.py:1229]: Can’t we just `self.demotables.discard(node.name)` here, just have `self.written_fields.add(node.name)` in the if below and get rid of `self.read_fields` and the discarding code above (around line 1214) entirely? Maybe I also missed something…

**Comment by gronerl** [comment]
Can you please provide an example where this PR changes the behavior? E.g. by adding a test. As I read the codes, they do exactly the same thing, it's just that instead of visiting the assign target, the respective code is done on the assign statement node directly.

File: src/gt4py/analysis/passes.py [hunk 2]
```diff
@@ -1185,6 +1185,10 @@ class DemoteLocalTemporariesToVariablesPass(TransformPass):
 
     This may occur because these can be local variables in the scope, and
     therefore do not need to be fields.
+
+    A field can be demoted when it is a temporary field that:
+    1. is never used with an offset
```
Comment (gronerl) [inline src/gt4py/analysis/passes.py:1190]:
```suggestion
    1. is never used with an offset and
```

**Comment by jdahm** [comment]
@gronerl You're right - when I first opened the PR there were differences in behavior, but it was breaking some of the FV3 stencils. Now that that is fixed, the behavior is the same as the existing code, so this is not really needed.

This branch is incorporated into the `horizontal-regions` branch where it is extended with a check on `HorizontalIf` nodes.

I will let you decide if we should merge this logic cleanup, or wait to merge as part of the `horizontal-regions` work.

**Comment by jdahm** [comment]
Closing for now.

**Pull request closed without merging**